crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,176 @@
+///   to send the data to the driver by writing it into the buffers of a receive queue descriptor
+///   chain.
+
+/// Console device errors.
+#[derive(Debug)]
+pub enum Error {
+    /// Failed to write slice to guest memory.
+    WriteToGuestFailed(GuestMemoryError),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            WriteToGuestFailed(ref err) => {
+                write!(f, "console failed to write slice to guest memory: {}", err)
+            }
+        }
+    }
+}
+
+/// Console struct that implements the abstraction of virtio console descriptor chain handling.
+pub struct Console {
+    /// Buffer that stores data to be sent to the driver.
+    input_buffer: Arc<Mutex<Vec<u8>>>,
+}
+
+impl Console {
+    /// Create new console object.
+    pub fn new() -> Self {
+        Console {
+            input_buffer: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    /// Adds data in the form of raw bytes to the input buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - A mutable reference to a `Vec<u8>` containing the raw bytes that will be
+    ///            sent to the driver when the `process_receiveq_chian` method is called.
+    pub fn enqueue_data(&self, data: &mut Vec<u8>) {
+        let mut input_buffer = self.input_buffer.lock().unwrap();
+        input_buffer.append(data);
+    }
+
+    /// Checks if input buffer is empty.
+    pub fn is_input_buffer_empty(&self) -> bool {
+        let input_buffer = self.input_buffer.lock().unwrap();
+        input_buffer.is_empty()
+    }
+
+    /// Read data from a `desc_chain` of the transmit queue and write it to a specified output.
+    ///
+    /// # Arguments
+    /// * `desc_chain` - A mutable reference to the descriptor chain that should point to the
+    ///                  buffers that the virtio console driver has sent to the device.
+    /// * `output` - An object that implements the `Write` trait and specifies where the data
+    ///              should be written to (usually `stdout`). The `output` object is not included
+    ///              as part of the [`Console`](struct.Console.html) struct and rather passed as an
+    ///              argument to enable more flexibility in choosing and changing the output sink
+    ///              from a higher level without having to reinitialize the
+    ///              [`Console`](struct.Console.html) structure.
+    ///
+    /// # Returns
+    /// The length in bytes of the processes data.
+    pub fn process_transmitq_chain<M, T>(
+        &self,
+        desc_chain: &mut DescriptorChain<M>,
+        output: &mut T,
+    ) -> usize
+    where
+        M: Deref,
+        M::Target: GuestMemory,
+        T: Write,
+    {
+        // Iterate through all the buffers in the chain and write the data to output.
+        let mut len = 0;
+        while let Some(desc) = desc_chain.next() {
+            len += desc_chain
```

Comment:
The operation here is unchecked in the sense that we do not check whether len + write_len still fits in a `usize`. There are 2 possibilities here:
- either it's impossible for the operation to overflow, in which case we should write a comment about why is that the case, or
- the operation can overflow. In this case we have multiple options as well. We can either use `checked_add` and return an error when the overflow happens (reference for checked operations: https://doc.rust-lang.org/std/primitive.usize.html#method.checked_add), or we can make len a type that wraps around, or we can use [saturating_add](https://doc.rust-lang.org/std/primitive.usize.html#method.saturating_add). I think the best option would be to return an error in case of overflow.